src/object_store/src/object/opendal_engine.rs

```diff
@@ -0,0 +1,433 @@
+// Copyright 2023 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Cursor;
+
+use bytes::{BufMut, Bytes, BytesMut};
+use fail::fail_point;
+use futures::future::try_join_all;
+use futures::StreamExt;
+use itertools::Itertools;
+use opendal::services::{hdfs, memory};
+use opendal::Operator;
+use tokio::io::AsyncRead;
+
+use super::{
+    BlockLocation, BoxedStreamingUploader, ObjectError, ObjectMetadata, ObjectResult, ObjectStore,
+    StreamingUploader,
+};
+/// Opendal object storage.
+#[derive(Clone)]
+pub struct OpendalObjectStore {
+    op: Operator,
+    engine_type: EngineType,
+}
+#[derive(Clone)]
+enum EngineType {
+    Memory,
```

Comment:
I think it's better to keep both our in-memory object store and `opandal` memory store because
(1) There are some interfaces that are not implemented in `opandal` memory store, such as `delete objects`, `multi part`
(2) We need some unit tests to make sure the implementation of `impl ObjectStore for OpendalObjectStore` is correct, that is what `opandal` memory store does.